### ISSUE TITLE
IndexedVector and AffExpr

### ISSUE THREAD

**Issue by mlubin** [JuliaOpt/JuMP.jl#340]
I'm trying to use an ``IndexedVector`` where the coefficients are ``AffExpr`` but this doesn't quite work with the ``if val != zero(T)`` test which unintentionally creates a constraint object.
@IainNZ, have you run into this for JuMPeR?

**Comment by joehuchette** [comment]
Here's a crazy idea: what about removing all the methods for ``==`` and friends to construct constraints, and force users to use the macros? I don't see any reason why you would need to use the function version of ``addConstraint`` anymore, and there are all kinds of weird issues that crop up like this with having equality overloaded.

**Comment by mlubin** [comment]
Hmm. That could work for JuMP and ``AffExpr`` but in extensions we may still need to use operator overloading for constraints, at least until it's easy to extend the macros.

**Comment by mlubin** [comment]
I like the idea though.

**Comment by IainNZ** [comment]
I haven't encountered this in JuMPeR because I haven't tried - the places I would use it I just don't (e.g. printing).

Macros seem to work 100% with JuMPeR (although perf is ??? but I don't care) so I think I'd be OK with that too. Something to think about. Can we replace the equality test with `isequal` or something?

**Comment by IainNZ** [comment]
This reminds me of my PR to Julia relating to sparse matrices - someone suggested creating a new `Base.iszero` for this reason.

**Comment by joehuchette** [comment]
@IainNZ it looks like you only had to add [this](https://github.com/IainNZ/JuMPeR.jl/blob/05a398b678b07245945356d92e63886feefaf8ed/src/robustmacro.jl#L105-L109) to JuMPeR to make macros work, is that right? We probably make the code in JuMP more general to allow packages to use macros without touching the macro code themselves.

**Comment by joehuchette** [comment]
In particular, can't we just change the ``addToExpression`` fallback to the one in JuMPeR?

**Comment by mlubin** [comment]
In JuMPChance we need keyword arguments to ``addConstraint``:
```
addConstraint(m, z*x <= -1, with_probability=0.05,
    uncertainty_budget_mean=1, uncertainty_budget_variance=1)
```
I guess the macros could accept these and pass them along also.

**Comment by mlubin** [comment]
Okay, I'm coming around to this. In 0.3 we didn't want to provide the general fallback because it would enable slow multiplication of variable*coefficient. Now that that's resolved for the base case in JuMP, it makes sense to have the general fallback so that everything "just works" for extensions. The macro code will still need to be rearranged a bit to allow for the output expression type to be general, and we'll have to see if there are any performance hits from that.

**Comment by joehuchette** [comment]
We'll need to be careful, and have a long deprecation period, but I think I like it as well. How would you resolve the kwarg issue?

**Comment by mlubin** [comment]
I don't think it will be too difficult to parse them correctly and pass them onto ``addConstraint()``.

**Comment by IainNZ** [comment]
Ooh shoot I just realized I'd like/need kwargs too for JuMPeR!

**Comment by mlubin** [comment]
Resolved by #401
